src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs

```diff
@@ -298,6 +306,57 @@ public override SingleNamespaceOrTypeDeclaration VisitNamespaceDeclaration(Names
             }
 
             var diagnostics = DiagnosticBag.GetInstance();
+
+            if (node is SingleLineNamespaceDeclarationSyntax)
+            {
+                if (node.Parent is SingleLineNamespaceDeclarationSyntax)
+                {
+                    // Happens when user writes:
+                    //      namespace A.B;
+                    //      namespace X.Y;
+                    diagnostics.Add(ErrorCode.ERR_MultipleSingleLineNamespace, node.Name.GetLocation());
+                }
+                else if (node.Parent is NamespaceDeclarationSyntax)
+                {
+                    // Happens with:
+                    //
+                    //      namespace A.B
+                    //      {
+                    //          namespace X.Y;
+                    diagnostics.Add(ErrorCode.ERR_SingleLineAndNormalNamespace, node.Name.GetLocation());
+                }
+                else
+                {
+                    // Happens with cases like:
+                    //
+                    //      namespace A.B { }
+                    //      namespace X.Y;
+                    //
+                    // or even
+                    //
+                    //      class C { }
+                    //      namespace X.Y;
+
+                    Debug.Assert(node.Parent is CompilationUnitSyntax);
+                    var compilationUnit = (CompilationUnitSyntax)node.Parent;
+                    if (node != compilationUnit.Members[0])
+                    {
+                        diagnostics.Add(ErrorCode.ERR_SingleLineNamespaceNotBeforeAllMembers, node.Name.GetLocation());
+                    }
+                }
+            }
+            else
+            {
+                Debug.Assert(node is NamespaceDeclarationSyntax);
+
+                //      namespace X.Y;
+                //      namespace A.B { }
+                if (node.Parent is SingleLineNamespaceDeclarationSyntax)
+                {
+                    diagnostics.Add(ErrorCode.ERR_SingleLineAndNormalNamespace, node.Name.GetLocation());
+                }
+            }
```

Comment:
We should also have a check to disallow top-level statements.

Reply:
@RikkiGibson  Add PROTOTYPE or track elsewehere?

Reply:
I think this is handled by the existing mechanisms that prevent top-level statements within namespaces. See `SingleLineNamespaceWithFollowingStatement` in `AssemblyAndNamespaceTests.cs`